### PR TITLE
Add force change password toggle in portal's user detail

### PR DESCRIPTION
**Pull request by IniZio**
ref DEV-1627

dependencies:
- [x] #4503
- [x] #4511

### Screens

Screen|Shot
-|-
Default|image
Disabled when already expired|image
Toggle confirmation|image
Audit log|Screenshot 2024-07-31 at 11-00-07 Authgear Portal

### CODE REVIEW

**Comment by louischan-oursky** [comment]
@IniZio Please tag me next time, otherwise I may miss your PRs.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
Please rebase first.

**Comment by IniZio** [comment]
Rebased.

File: pkg/admin/facade/user.go [hunk 1]
```diff
@@ -109,6 +116,60 @@ func (f *UserFacade) ResetPassword(id string, password string, sendPassword bool
...
+		[]string{id},
+		&passwordType,
+		&primaryKind,
+	)
+	if err != nil {
+		return err
+	}
+
+	if len(ars) == 0 {
+		return api.ErrNoAuthenticator
```
Comment (louischan-oursky) [inline pkg/admin/facade/user.go:137]: According to the spec, we should introduce a new API error. See docs/specs/admin-api-change-password.md:85

File: pkg/admin/graphql/user_mutation.go [hunk 2]
```diff
@@ -164,6 +164,77 @@ var _ = registerMutationField(
 	},
 )
 
+var markPasswordAsExpiredInput = graphql.NewInputObject(graphql.InputObjectConfig{
+	Name: "MarkPasswordAsExpiredInput",
```
Comment (louischan-oursky) [inline pkg/admin/graphql/user_mutation.go]: According to the spec, the name is SetPasswordExpiredBlahBlahBlah

File: pkg/lib/authn/authenticator/password/provider.go [hunk 3]
```diff
@@ -95,13 +96,14 @@ type UpdatePasswordOptions struct {
...
-	}
 
 	newAuthen := a
-	if pwd.Compare([]byte(password), a.PasswordHash) != nil {
+	if options.SetPassword && pwd.Compare([]byte(password), a.PasswordHash) != nil {
+		err := p.PasswordChecker.ValidateNewPassword(a.UserID, password)
+		if err != nil {
+			return false, nil, err
+		}
+
```
Comment (louischan-oursky) [inline pkg/lib/authn/authenticator/password/provider.go:106]: The logic changed. It should be kept unchanged. We always validate the password first.

```golang
if options.SetPassword {
  err := p.PasswordChecker.ValidateNewPassword(a.UserID, password)
  if err != nil {
    return false, nil, err
  }
  if pwd.Compare([]byte(password), a.PasswordHash) != nil {
    newAuthen = p.populatePasswordHash(a, password)
  }
}
```

File: portal/src/graphql/adminapi/ChangePasswordScreen.tsx [hunk 4]
```diff
@@ -166,8 +168,9 @@ const ChangePasswordScreen: React.VFC = function ChangePasswordScreen() {
         state.sendPassword,
         state.setPasswordExpired
       );
+      await refetchUser();
```
Comment (louischan-oursky) [inline portal/src/graphql/adminapi/ChangePasswordScreen.tsx:171]: Do not do this. Instead, you should modify resetPasswordMutation.graphql to fetch the updated fields.

File: pkg/api/event/nonblocking/adminapi_mutation_markpasswordasexpired_executed.go [hunk 5]
```diff
@@ -6,44 +6,44 @@ import (
...
 const (
-	AdminAPIMutationMarkPasswordAsExpiredExecuted event.Type = "admin_api.mutation.mark_password_as_expired.executed" // nolint:gosec
+	AdminAPIMutationSetPasswordExpiredExecuted event.Type = "admin_api.mutation.set_password_expired.executed" // nolint:gosec
 )
 
-type AdminAPIMutationMarkPasswordAsExpiredExecutedEventPayload struct {
+type AdminAPIMutationSetPasswordExpiredExecutedEventPayload struct {
 	UserRef   model.UserRef `json:"-" resolve:"user"`
 	UserModel model.User    `json:"user"`
 	IsExpired bool          `json:"is_expired"`
```
Comment (louischan-oursky) [inline pkg/api/event/nonblocking/adminapi_mutation_markpasswordasexpired_executed.go:15]: Expired bool json:"expired"

File: portal/src/graphql/adminapi/schema.graphql [hunk 6]
```diff
@@ -1471,6 +1456,21 @@ type SetDisabledStatusPayload {
   user: User!
 }
 
+""""""
+input SetPasswordExpiredInput {
+  """Indicate whether the user's password is expired."""
+  isExpired: Boolean!
```
Comment (louischan-oursky) [inline portal/src/graphql/adminapi/schema.graphql:1462]: According to the spec, it is called `expired`

**Comment by IniZio** [comment]
`npm audit` is failing for authui, the fix requires major version bump for `ethers` to `v6.13.2`. 

Created https://linear.app/authgear/issue/DEV-1721/npm-audit-fails-for-authui-ethers6132 first.

File: pkg/admin/graphql/user_mutation.go (hunk 2 above)
Comment (IniZio) [reply pkg/admin/graphql/user_mutation.go]: It was updated in later commit.

File: portal/src/graphql/adminapi/schema.graphql (hunk 6 above)
Comment (IniZio) [reply portal/src/graphql/adminapi/schema.graphql]: Updated along with above comments :adore2x:

File: pkg/api/event/nonblocking/adminapi_mutation_markpasswordasexpired_executed.go [hunk 7]
```diff
@@ -12,7 +12,7 @@ const (
 type AdminAPIMutationSetPasswordExpiredExecutedEventPayload struct {
 	UserRef   model.UserRef `json:"-" resolve:"user"`
 	UserModel model.User    `json:"user"`
-	IsExpired bool          `json:"is_expired"`
+	Expired   bool          `json:"expire"`
```
Comment (louischan-oursky) [inline pkg/api/event/nonblocking/adminapi_mutation_markpasswordasexpired_executed.go:15]: typo expire -> expired

File: pkg/lib/infra/mail/sender.go [hunk 8]
```diff
@@ -61,16 +61,16 @@ func (s *Sender) Send(opts SendOptions) (err error) {
...
+	// if s.DevMode {
+	// 	s.Logger.
+	// 		WithField("recipient", opts.Recipient).
+	// 		WithField("body", opts.TextBody).
+	// 		WithField("sender", opts.Sender).
+	// 		WithField("subject", opts.Subject).
+	// 		WithField("reply_to", opts.ReplyTo).
+	// 		Warn("skip sending email in development mode")
+	// 	return nil
+	// }
```
Comment (louischan-oursky) [inline pkg/lib/infra/mail/sender.go:73]: Do not commit this.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
You can rebase on the latest main to have https://linear.app/authgear/issue/DEV-1721/npm-audit-fails-for-authui-ethers6132 fixed.

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**